### PR TITLE
[Numpy] Random.choice implemented

### PR DESCRIPTION
**Pull request by xidulu**
## Description ##
As title
https://docs.scipy.org/doc/numpy-1.16.1/reference/generated/numpy.random.choice.html

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Uniform/Weighted sampling with/without replacement implemented.
- [x] Allow input to be array/symbol. 

## Comments ##
- Weighted sampling has some performance issues currently, as TopK used in weighted sample without replacement is implemented using `sort`.

- Heavily referenced from https://github.com/pytorch/pytorch/pull/18624

### CODE REVIEW

File: python/mxnet/numpy/random.py [hunk 1]
```diff
@@ -180,3 +180,60 @@ def multinomial(n, pvals, size=None, **kwargs):
...
+        single value is returned.
+    replace : boolean, optional
+        Whether the sample is with or without replacement
+    p : 1-D array-like, optional
+        The probabilities associated with each entry in a.
+        If not given the sample assumes a uniform distribution over all
+        entries in a.
+    ctx : Context, optional
+        Device context of output. Default is current context.
+    out : ``ndarray``, optional
```
Comment (haojin2) [inline python/mxnet/numpy/random.py]: What's the usage of `out` here? I think you can get rid of it here.

File: src/operator/numpy/random/np_choice_op.h [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+          random_numbers.dptr_, outputs[0].dptr<int64_t>());
+    } else {
+      Tensor<xpu, 1, int64_t> indices = Tensor<xpu, 1, int64_t>(
+          reinterpret_cast<int64_t *>(workspace_ptr), Shape1(indices_size), s);
+      indices = expr::range((int64_t)0, input_size);
+      Kernel<generate_keys, xpu>::Launch(s, input_size, random_numbers.dptr_,
+                                         inputs[weight_index].dptr<float>());
+      _sort<xpu>(random_numbers.dptr_, indices.dptr_, input_size);
+      Copy(outputs[0].FlatTo1D<xpu, int64_t>(s), indices.Slice(0, output_size),
+           s);
```
Comment (haojin2) [inline src/operator/numpy/random/np_choice_op.h]: line length limit is 100 characters:
```c++
      Copy(outputs[0].FlatTo1D<xpu, int64_t>(s), indices.Slice(0, output_size), s);
```

File: src/operator/numpy/random/np_choice_op.h [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    }
+  } else {
+    Random<xpu, unsigned> *prnd = ctx.requested[0].get_random<xpu, unsigned>(s);
+    int64_t random_tensor_size =
+        (replace ? output_size
+                 : std::min(output_size, input_size - output_size));
+    int64_t indices_size = replace ? 0 : input_size;
+    Tensor<xpu, 1, char> workspace =
+        (ctx.requested[1].get_space_typed<xpu, 1, char>(
```
Comment (haojin2) [inline src/operator/numpy/random/np_choice_op.h]: why an extra pair of parentheses?

Comment (xidulu) [reply src/operator/numpy/random/np_choice_op.h]: Thx for pointing out, working on it

**Pull request merged**